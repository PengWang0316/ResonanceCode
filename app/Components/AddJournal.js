import React, { Component } from "react";
import Util from "../apis/Util";
import Loading from "./Loading"
import JournalContent from "./JournalContent";
import ReadingSearchAndList from "./ReadingSearchAndList";
import { isLogin } from "../apis/LoginApi";
import DatabaseApi from "../apis/DatabaseApi";
import PropTypes from "prop-types";

class AddJournal extends Component {

  componentWillMount(){
    // Initial variables
    this.readings = this.props.journalData ? this.props.journalData.readingIds : {}; // keep which readings should be attached on. Format is like {readingId: null}
    this.contents = {}; // Keep content keys and content
    this.contentKeyIndex = 0; // Generate keys for different contents
    this.contentIndexs = {}; // Use to track the index of content component in the array. Delete function needs it. The format is {contentKey: index}
    this.jounalId = this.props.journalData ? this.props.journalData._id : null; // Keeping journal id for update
    this.userId = isLogin(document).userid;

    this.state={
      isWriting: false,
      journalDate: this.props.journalData ? Util.getDateString(this.props.journalData.date) : Util.getCurrentDateString(),
      isDateCorrect: true,
      isEmptyReading: this.props.journalData ? false : true,
      contentComponentArray: [], // keep content component
      addJournalContent: "overview",
      pingPongState: this.props.journalData ? this.props.journalData.ping_pong_state : "Neutral"
    };

    console.log("Journal before alter: ", this.props.journalData);
    // If journal data exsit, get content
    if(this.props.journalData){
      // this.oldContentKeys = []; // keep original content keys
      this.oldReadingIds = Object.keys(this.readings); // keep original reading ids

      // delete unnecessary properties from journal object
      let journal = this.props.journalData;
      this.journalUserId = journal.user_id;
      delete journal.date;
      delete journal.user_id;
      delete journal.ping_pong_state;
      delete journal._id;
      delete journal.readingIds;
      let keyRegExp = /-(\d+)$/; // The regular expression for subtract suffixs
      console.log("Journal after alter: ", this.props.journalData);
      Object.keys(journal).map((key)=>{
        // this.oldContentKeys.push(key); // keeping the exsit keys in an array for delete function
        // handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared)
        // Also have to save the largest key number and use it to countine
        // journalData property is like {'overview-0': 'test overview', 'overview-0-isShared': true}
        // if key is isShared do not do anything
        let matchResult = key.match(keyRegExp);
        if (matchResult){
          if (this.contentKeyIndex < matchResult[1]) this.contentKeyIndex = matchResult[1]; // keep the max index number and user can contiune add new content
          this.handleAddContentClick(journal[key], key.replace(keyRegExp, ""), key, matchResult[1], journal[`${key}-isShared`]);
          // console.log("isShared: ",journal[`${key}-isShared`]);
        }
      });
      this.contentKeyIndex++;
    }

  }

  handleChange(event, element){
    let newState={};
    newState[element]=event.target.value;
    if(element=="journalDate") newState.isDateCorrect = Util.matchDateFormat(newState[element]);
    this.setState(newState);
  }

  handleChangeCallback(contentKey, contentText){
    console.log(contentKey, contentText);
    // update content in state
    this.contents[contentKey]=contentText;
  }

  handleDeleteContentCallback(contentKey){
    // Remove content from state
    if(this.contents.hasOwnProperty(contentKey)){
      delete this.contents[contentKey];
      delete this.contents[`${contentKey}-isShared`];
    }
    // Remove content component
    delete this.contentIndexs[contentKey];
    this.setComponentToStateArray();
    // this.state.contentComponentArray.splice(this.contentIndexs[contentKey],1);
    // this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSharedBoxChangeCallback(contentKey, isShared){
    this.contents[`${contentKey}-isShared`] = isShared;
  }

  handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared){
    // put component in array in order to show
    newContentKey = newContentKey ? newContentKey : `${this.state.addJournalContent}-${this.contentKeyIndex}`;
    // put the component in the track object
    this.contentIndexs[newContentKey]=<JournalContent key={contentKeyIndex ? contentKeyIndex : this.contentKeyIndex++} newContent={addJournalContent ? addJournalContent : ""} newContentName={newContentName ? newContentName : this.state.addJournalContent} newContentKey={newContentKey} handleChangeCallback={(contentKey, contentText)=>{this.handleChangeCallback(contentKey, contentText);}} handleDeleteContentCallback={(contentKey)=>{this.handleDeleteContentCallback(contentKey);}} handleSharedBoxChangeCallback={(contentKey, isShared)=>{this.handleSharedBoxChangeCallback(contentKey, isShared)}} isShared={isShared ? true : false} />;
    this.setComponentToStateArray();
    // this.setState({contentComponentArray: this.state.contentComponentArray});

    // also have to put content title to state in order to track content and update
    this.contents[newContentKey] = addJournalContent ? addJournalContent : "";
    this.contents[`${newContentKey}-isShared`] = isShared ? true : false;
  }

  setComponentToStateArray(){
    // transfor object to array in order to display
    this.state.contentComponentArray=[];
    Object.keys(this.contentIndexs).map((element)=>{
      this.state.contentComponentArray.push(this.contentIndexs[element]);
    });
    this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSubmit(event, isUpdate){
    event.preventDefault();
    // console.log("submit: contents:",this.contents);
    console.log("journal date:",this.state.journalDate);
    // Assemble a journal object for save
    let readingIdArray = [];
    Object.keys(this.readings).map((key)=>{readingIdArray.push(key)});
    let journal = Object.assign({_id:this.jounalId, date: new Date(this.state.journalDate), ping_pong_state: this.state.pingPongState, user_id: this.userId, readingIds: readingIdArray}, this.contents);

    if(isUpdate){
      console.log("submit: journal:", journal);
      // console.log("oldContentKeys:", this.oldContentKeys);
      console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      let deleteContents = [];
      let deleteReadingIds = [];
      // this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      this.oldReadingIds.map((element)=>{if(!this.readings.hasOwnProperty(element)) deleteReadingIds.push(element);});
      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      console.log("delete contents: ",deleteContents);
      console.log("delete reading ids : ",deleteReadingIds);
      DatabaseApi.updateJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }else{
      DatabaseApi.createJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }
    /**/
    this.setState({isWriting: true});
    console.log("submit: journal:", journal);
  }

  /* For ReadingSearchAndList callback */
  handleAttachReadingCallback(readingId){
    this.readings[readingId] = null;
    this.setState({isEmptyReading: false});
  }

  handleDetachAttachReadingCallback(readingId){
    delete this.readings[readingId];
    this.setState({isEmptyReading: Object.keys(this.readings).length===0});
  }

  handleCancel(){
    this.props.history.push("/reading");
  }

  handleDelete(){
    console.log("Delete journal!");
  }

  render(){
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        <div className="titleDiv">{this.props.journalData ? "Update journal for readings" : "Add a new journal for readings"}</div>
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event, this.props.journalData);}}>

          <div className="text-right bottom-btn-div">
            {(!this.props.journalData || this.userId == this.journalUserId) &&
              <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
            }
            {(this.props.journalData && this.userId == this.journalUserId) &&
              <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
            }
            <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
          </div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">

                <input className={this.state.isDateCorrect?"form-control":"form-control form-control-warning"} type="text" placeholder="mm/dd/yyyy" id="journalDate" value={this.state.journalDate} onChange={(event)=>{this.handleChange(event, "journalDate")}} />
                {!this.state.isDateCorrect && <span className="glyphicon glyphicon-warning-sign form-control-feedback form-control-warning-span"></span>}

              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            {/* New content goes here */}
            {this.state.contentComponentArray}

            {/* Add content button and drop list */}
            <div className="row addJournalContentDiv">

              <div onClick={()=>{this.handleAddContentClick();}} className="addJournalContentBtnDiv col-sm-6"><i className="fa fa-plus-square" /> Add one content for your journal</div>
              <div className="col-sm-6">
                <select  className="form-control" onChange={(event)=>{this.handleChange(event,"addJournalContent");}}>
                  <option value="overview">Overview</option>
                  <option value="internal_environment">Internal Environment</option>
                  <option value="relational_environment">Relational Environment</option>
                  <option value="physical_environment">Physical Environment</option>
                  <option value="creative_vector">Creative Vector</option>
                  <option value="synchronicities">Synchronicities</option>
                  <option value="dreams">Dreams</option>
                  <option value="other">Other</option>
                </select>

              </div>

            </div>


            {/*  Start reading search function */}
            <ReadingSearchAndList readings={this.props.journalData ? this.readings : null} attachReadingCallback={(readingId)=>{this.handleAttachReadingCallback(readingId);}} detachReadingCallback={(readingId)=>{this.handleDetachAttachReadingCallback(readingId);}} />

              <div className="text-right bottom-btn-div">
                {(!this.props.journalData || this.userId == this.journalUserId) &&
                  <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
                }
                {(this.props.journalData && this.userId == this.journalUserId) &&
                  <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
                }
                <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
              </div>
          </form>

        </div>
    );
  }


}
AddJournal.propTypes={
  journalData: PropTypes.object,
  history: PropTypes.object
};
export default AddJournal;
