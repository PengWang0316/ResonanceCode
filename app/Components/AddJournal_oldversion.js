import React, { Component } from "react";
import QueryString from "query-string";
import PropTypes from "prop-types";
import Util from "../apis/Util";
// import api from "../apis/api";
import DatabaseApi from "../apis/DatabaseApi";
import Loading from "./Loading";
import LoginApi from "../apis/LoginApi";

class AddJournal_oldversion extends Component{

  componentWillMount(){

    if(this.props.location){
      let queryInfo = QueryString.parse(this.props.location.search);
      this.readingId = queryInfo.readingId;
      this.readingName = queryInfo.readingName;
      this.readingDate = Util.getDateString(queryInfo.readingDate);
      // console.log("AddJournal paper readingId, readingName, and readingDate:", `Id: ${this.readingId} Name: ${this.readingName} Date: ${this.readingDate}`);
    }
    this.userId=LoginApi.isLogin(document).userid;

    let journalData = this.props.journalData;
    if(journalData){
      this.journalId=journalData.journalId;
      // this.readingDate=journalData.readingDate;
      this.readingName=journalData.readingName;
    }
    // console.log("AddJournal page:",journalData);
    // setting state
    this.state={
      isWriting: false,
      journalDate: journalData?Util.getDateString(journalData.date):Util.getCurrentDateString(),
      pingPongState: journalData?journalData.pingPongState:"Neutral",
      overview: journalData?journalData.overview:"",
      internalEnvironment: journalData?journalData.internalEnvironment:"",
      relationalEnvironment: journalData?journalData.relationalEnvironment:"",
      physicalEnvironment: journalData?journalData.physicalEnvironment:"",
      creativeVector: journalData?journalData.creativeVector:"",
      synchronicities: journalData?journalData.synchronicities:"",
      dreams: journalData?journalData.dreams:"",
      journalId: journalData?journalData.journalId:"",
      userId: journalData?journalData.userId:"",
      readingId: journalData?journalData.readingId:""
    };
    // if(this.journalId) this.setState(this.state);
  }

  componentDidMount(){
    // console.log("did mount");
    if(this.journalId) this.setState(this.state);
  }

  handleChange(event, element){
    // console.log(event.target.value);
    let newState={};
    newState[element]=event.target.value;
    this.setState(newState);
    // console.log(newState);
  }

  handleSubmit(event){
    event.preventDefault();
    // console.log("**********submit************");

      this.setState({isWriting: true});
      // console.log("submit:",this.state);
      // assemble object for creating
      let journalObject = {
        user_id: this.userid,
        reading_id: this.readingId?this.readingId:this.state.readingId,
        date: new Date(this.state.journalDate),
        overview: this.state.overview,
        internal_environment: this.state.internalEnvironment,
        relational_environment: this.state.relationalEnvironment,
        physical_environment: this.state.physicalEnvironment,
        creative_vector: this.state.creativeVector,
        synchronicities: this.state.synchronicities,
        dreams: this.state.dreams,
        ping_pong_state:this.state.pingPongState
      };
      // console.log("journalId", this.journalId);
      if(this.journalId) DatabaseApi.updateJournal(journalObject, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
      else DatabaseApi.createJournal(journalObject).then((reault)=>{this.props.history.push("/reading")});

  }

  handleDelete(event){
    event.preventDefault();
    // console.log("**********delete************");
    DatabaseApi.deleteJournal(this.state.readingId, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
  }

  render(){
    let titleDiv;
    if (this.readingId) titleDiv = <div><span className="rcTitle">Add a new journal to your reading: </span>{this.readingName}  ({this.readingDate})</div>;
      else titleDiv = <div><span className="rcTitle">Update {this.readingName}</span></div>;
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        {titleDiv}
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event);}}>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button type="button" className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">
                <input className="form-control" type="text" value={this.state.journalDate} id="journalDate" onChange={(event)=>{this.handleChange(event,"journalDate");}} />
              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="overview" className="col-form-label">Overview</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.overview} placeholder="Overview..." id="overview" onChange={(event)=>{this.handleChange(event,"overview");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="internalEnvironment" className="col-form-label">Internal Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.internalEnvironment} placeholder="Internal Environment..." id="internalEnvironment" onChange={(event)=>{this.handleChange(event,"internalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="relationalEnvironment" className="col-form-label">Relational Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.relationalEnvironment} placeholder="Relational Environment..." id="relationalEnvironment" onChange={(event)=>{this.handleChange(event,"relationalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="physicalEnvironment" className="col-form-label">Physical Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.physicalEnvironment} placeholder="Physical Environment..." id="physicalEnvironment" onChange={(event)=>{this.handleChange(event,"physicalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="creativeVector" className="col-form-label">Creative Vector</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.creativeVector} placeholder="Creative Vector..." id="creativeVector" onChange={(event)=>{this.handleChange(event,"creativeVector");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="synchronicities" className="col-form-label">Synchronicities</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.synchronicities} placeholder="Synchronicities..." id="synchronicities" onChange={(event)=>{this.handleChange(event,"synchronicities");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="dreams" className="col-form-label">Dreams</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.dreams} placeholder="Dreams..." id="dreams" onChange={(event)=>{this.handleChange(event,"dreams");}} />
              </div>
            </div>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

        </form>

      </div>
    );
  }

}
AddJournal_oldversion.propTypes={
  journalData: PropTypes.object
};
export default AddJournal_oldversion;
