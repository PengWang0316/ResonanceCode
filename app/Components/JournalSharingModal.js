import React, { Component } from 'react';
import { connect } from 'react-redux';

import '../styles/JournalSharingModal.module.css';
import { NUMBER_OF_USER_PER_PAGE } from '../config';
import { fetchUsersAmount, fetchAllUserList } from '../actions/UserActions';
import { updateJournalShareList } from '../actions/JournalActions';
import Pagination from './SharedComponents/Pagination';
import LoadingAnimation from './SharedComponents/LoadingAnimation';

/** Show the journal sharing modal */
export class JournalSharingModal extends Component {
  state = {
    shareList: {}
  };
  /** Fetching the total amount number of user and the first page user name.
    * @return {null} No return.
  */
  componentWillMount() {
    this.props.fetchUsersAmount();
    this.props.fetchAllUserList(0);
  }

  /** When receive a new journal from props, save its sharingList to state.
    * @param {object} nextProps is an object that contains props' value.
    * @return {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if ((this.props.journal !== nextProps.journal) &&
    nextProps.journal) {
      const shareList = {};
      if (nextProps.journal.shareList)
        nextProps.journal.shareList.forEach(share => { shareList[share.id] = share; });
      this.setState({ shareList });
    }
  }

  /** Removing the user from state when the user click remove button.
    * @param {object} target is the html element a user is clicking.
    * @return {null} No return.
  */
  handleRemoveUser = ({ target }) => {
    const newShareList = Object.assign({}, this.state.shareList);
    delete newShareList[target.getAttribute('userid')];
    this.setState({ shareList: newShareList });
  };

  /** Adding the user to the state when a user click a user's name from user name list. The html element includes a img or i tag and a text element.
    * @param {object} target is the html element a user is clicking.
    * @return {null} No return.
  */
  handleAddUser = ({ target }) => this.setState({
    shareList: Object.assign({
      [target.getAttribute('userid')]: {
        id: target.getAttribute('userid'),
        displayName: target.innerText,
        photo: target.childNodes[0].getAttribute('src') ? target.childNodes[0].getAttribute('src') : '',
        sharedDate: new Date()
      }
    }, this.state.shareList)
  });

  /** Saving the sharing information back to the journal.
    * @return {null} No return.
  */
  handleSave = () => {
    this.props.updateJournalShareList({
      journalId: this.props.journal._id,
      readingId: this.props.readingId,
      shareList: Object.keys(this.state.shareList).map(key => this.state.shareList[key])
    });
    $('#journalSharingModal').modal('toggle'); // $ will use the jQuery libaray from index.html.
  };

  /** Rendering the jsx for the component.
    * @return {jsx} Returning the jsx for the component.
  */
  render() {
    const { users, usersAmount } = this.props;
    return (
      <div className="modal fade" id="journalSharingModal" tabIndex="-1" role="dialog" aria-labelledby="journalSharingModalLabel" aria-hidden="true">
        <div className="modal-dialog" role="document">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title" id="journalSharingModalLabel">Share this journal with others</h5>
              <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                <span aria-hidden="true">&times;</span>
              </button>
            </div>
            <div className="modal-body">
              <div><b>Sharing List</b></div>
              <div className="text-muted">This journal will be shared with users are listed below.</div>
              <div className="shareListDiv d-flex flex-wrap mb-4">
                {this.state.shareList && Object.keys(this.state.shareList).map(key => (
                  <div key={this.state.shareList[key].id} className="mr-3">
                    {this.state.shareList[key].photo ? <img className="avatar-photo" src={this.state.shareList[key].photo} alt="user" /> : <i className="fa fa-user-circle" aria-hidden="true" />} {this.state.shareList[key].displayName} <i userid={this.state.shareList[key].id} role="button" className="fa fa-times closeBtn" aria-hidden="true" title={`Stop sharing with ${this.state.shareList[key].displayName}`} onClick={this.handleRemoveUser} />
                  </div>
                ))}
              </div>
              {/* Starting to show user searching component */}
              <div>
                <div className="mb-2"><b>Pick users you want to share this journal with</b></div>
                <div>
                  <form className="form-inline">
                    <div className="form-group mr-2 mb-0">
                      <label htmlFor="inputUsername" className="sr-only">User Name</label>
                      <input type="password" className="form-control" id="inputUsername" placeholder="User's name" />
                    </div>
                    <button type="submit" className="btn btn-primary btn-sm">Search</button>
                  </form>
                </div>
                <div className="userListDiv d-flex flex-wrap mt-3">
                  <LoadingAnimation />
                  {users && users.map(user => {
                    if (Object.prototype.hasOwnProperty.call(this.state.shareList, user._id) ||
                    user._id === this.props.user._id) return null;
                    return <div role="button" tabIndex="-1" onClick={this.handleAddUser} userid={user._id} key={user._id} className="userNameDiv">{user.photo ? <img className="avatar-photo mr-2" src={user.photo} alt="user" /> : <i className="fa fa-user-circle mr-2" aria-hidden="true" />}{user.displayName}</div>;
                  })
                  }
                  {usersAmount !== null &&
                    <div className="mt-3 w-100 d-flex justify-content-end">
                      <Pagination
                        amount={usersAmount}
                        fetchContent={fetchAllUserList}
                        numberPerpage={NUMBER_OF_USER_PER_PAGE}
                      />
                    </div>}
                </div>
              </div>
            </div>
            <div className="modal-footer">
              <button type="button" className="btn btn-secondary" data-dismiss="modal">Close</button>
              <button type="button" className="btn btn-primary" onClick={this.handleSave}>Save</button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  journal: state.journal,
  users: state.users,
  usersAmount: state.usersAmount,
  user: state.user
});
const mapDispatchToProps = dispatch => ({
  fetchUsersAmount: _ => dispatch(fetchUsersAmount()),
  fetchAllUserList: _ => dispatch(fetchAllUserList()),
  updateJournalShareList: params => dispatch(updateJournalShareList(params))
});
export default connect(mapStateToProps, mapDispatchToProps)(JournalSharingModal);
