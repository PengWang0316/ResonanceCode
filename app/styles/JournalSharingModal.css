.shareListDiv, .userListDiv{
  /*max-width: 400px;*/
  min-height: 150px;
  border-radius: 5px;
}
.shareListDiv{
  border: 1px solid #e2e2e2;
  padding: 8px;
}
.userNameDiv{
  background: #a1cff0;
  color: #FFF;
  border-radius: 4px;
  padding: 8px;
  margin: 5px;
  -webkit-transition: background .25s ease .25s;
  -moz-transition: background .25s ease .25s;
  transition: background .25s ease .25s;
}
.userNameDiv:hover{
  background: #4b96cc;
  cursor: pointer;
}
.closeBtn{
  -webkit-transition: color .25s;
  -moz-transition: color .25s;
  transition: color .25s;
}
.closeBtn:hover{
  cursor: pointer;
  color: #a31d1d;
}
