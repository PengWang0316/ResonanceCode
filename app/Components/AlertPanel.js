import React from 'react';

import styles from '../styles/AlertPanel.module.css';

const AlertPanel = ({
  id = '', type = 'success', children = '', style = {}
}) => (
  <div id={id} className={`alert alert-${type} mt-3 ${styles.alertPanel}`} style={style} role="alert">
    {children}
  </div>
);
export default AlertPanel;
