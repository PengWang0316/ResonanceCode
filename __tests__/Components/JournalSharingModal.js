import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { JournalSharingModal } from '../../app/Components/JournalSharingModal';

jest.mock('../../app/Components/UserPicker', () => 'UserPicker');

describe('JournalSharingModal test', () => {
  sinon.useFakeTimers(new Date(2018, 0, 1));
  const defaultProps = {
    journal: { _id: 'journalId' },
    users: [{}],
    usersAmount: 5,
    user: {
      _id: 'userId',
      settings: {
        userGroups: {
          'Group A': [
            {
              id: '59f2b6f5ed9baa00546daf8d',
              displayName: 'testnam',
              photo: 'photoA'
            },
            {
              id: '59f39981ad0e48008bffc202',
              displayName: 'usertestuser',
              photo: 'photoB'
            }
          ]
        }
      }
    },
    readingId: 'readingId',
    fetchUsersAmount: jest.fn(),
    fetchAllUserList: jest.fn(),
    updateJournalShareList: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<JournalSharingModal {...props} />);

  test('initial functions called', () => {
    getShallowComponent();
    expect(defaultProps.fetchUsersAmount).toHaveBeenCalledTimes(1);
    expect(defaultProps.fetchAllUserList).toHaveBeenLastCalledWith(0);
  });

  test('componentWillReceiveProps', () => {
    const component = getShallowComponent({ ...defaultProps, journal: {} });
    component.setProps({ journal: { shareList: [{ id: 1 }, { id: 2 }] } });
    expect(component.instance().existedShareList).toEqual([1, 2]);
    expect(Object.keys(component.state('shareList')).length).toBe(2);
  });

  test('handleRemoveUserCallback', () => {
    const component = getShallowComponent();
    component.setState({ shareList: { idA: {}, idB: {} } });
    const mockFunction = jest.fn().mockReturnValue('idA');
    component.find('UserPicker').prop('handleRemoveUser')({ target: { getAttribute: mockFunction } });
    expect(mockFunction).toHaveBeenLastCalledWith('userid');
    expect(component.state('shareList').idA).toBeUndefined();
  });

  test('handleAddUserCallback target has childNodes', () => {
    const component = getShallowComponent();
    const mockGetAttr = jest.fn().mockReturnValue('imageSrc');
    const mockTargeGetAttr = jest.fn().mockReturnValue('newUserId');
    component.find('UserPicker').prop('handleAddUser')({ target: { childNodes: [{ getAttribute: mockGetAttr }], getAttribute: mockTargeGetAttr, innerText: 'innerText' } });
    expect(mockGetAttr).toHaveBeenLastCalledWith('src');
    expect(mockTargeGetAttr).toHaveBeenLastCalledWith('userid');
    expect(component.state('shareList').newUserId).toEqual({
      id: 'newUserId', displayName: 'innerText', photo: 'imageSrc', sharedDate: new Date()
    });
  });

  test('handleAddUserCallback target without childNodes', () => {
    const component = getShallowComponent();
    const mockTargetGetAttr = jest.fn().mockReturnValue('imageSrc');
    const mockParentGetAttr = jest.fn().mockReturnValue('newUserId');
    component.find('UserPicker').prop('handleAddUser')({
      target: {
        getAttribute: mockTargetGetAttr, parentElement: { getAttribute: mockParentGetAttr }, childNodes: [], innerText: 'innerText'
      }
    });
    expect(mockTargetGetAttr).toHaveBeenLastCalledWith('src');
    expect(mockParentGetAttr).toHaveBeenLastCalledWith('userid');
    expect(component.state('shareList').newUserId).toEqual({
      id: 'newUserId', displayName: 'innerText', photo: 'imageSrc', sharedDate: new Date()
    });
  });

  test('handleGroupClickCallback', () => {
    const component = getShallowComponent();
    component.find('UserPicker').prop('handleGroupClick')({ target: { id: 'Group A' } });
    const shareList = component.state('shareList');
    expect(shareList['59f2b6f5ed9baa00546daf8d']).toEqual({
      id: '59f2b6f5ed9baa00546daf8d', displayName: 'testnam', photo: 'photoA', sharedDate: new Date()
    });
    expect(shareList['59f39981ad0e48008bffc202']).toEqual({
      id: '59f39981ad0e48008bffc202', displayName: 'usertestuser', photo: 'photoB', sharedDate: new Date()
    });
  });

  test('handleSave', () => {
    const component = getShallowComponent();
    const mockModalFunc = jest.fn();
    window.$ = jest.fn().mockReturnValue({ modal: mockModalFunc });
    component.find('button').at(2).simulate('click');
    expect(defaultProps.updateJournalShareList).toHaveBeenCalledTimes(1);
    expect(window.$).toHaveBeenLastCalledWith('#journalSharingModal');
    expect(mockModalFunc).toHaveBeenLastCalledWith('toggle');
  });

  test('JournalSharingModal snapshot', () => expect(renderer.create(<JournalSharingModal {...defaultProps} />).toJSON()).toMatchSnapshot());
});
