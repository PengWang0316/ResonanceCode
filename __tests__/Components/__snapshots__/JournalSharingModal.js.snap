// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalSharingModal test JournalSharingModal snapshot 1`] = `
<div
  aria-hidden="true"
  aria-labelledby="journalSharingModalLabel"
  className="modal fade"
  id="journalSharingModal"
  role="dialog"
  tabIndex="-1"
>
  <div
    className="modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h5
          className="modal-title"
          id="journalSharingModalLabel"
        >
          Share this journal with others
        </h5>
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
      <div
        className="modal-body"
      >
        <div>
          <b>
            Sharing List
          </b>
        </div>
        <div
          className="text-muted"
        >
          You are sharing this journal with users listed below.
        </div>
        <UserPicker
          chooseTextTitle="Pick users you want to share this journal with"
          fetchAllUserList={[Function]}
          handleAddUser={[Function]}
          handleGroupClick={[Function]}
          handleRemoveUser={[Function]}
          removeText="Stop sharing with"
          user={
            Object {
              "_id": "userId",
              "settings": Object {
                "userGroups": Object {
                  "Group A": Array [
                    Object {
                      "displayName": "testnam",
                      "id": "59f2b6f5ed9baa00546daf8d",
                      "photo": "photoA",
                    },
                    Object {
                      "displayName": "usertestuser",
                      "id": "59f39981ad0e48008bffc202",
                      "photo": "photoB",
                    },
                  ],
                },
              },
            }
          }
          userList={Object {}}
          users={
            Array [
              Object {},
            ]
          }
          usersAmount={5}
        />
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-secondary"
          data-dismiss="modal"
          type="button"
        >
          Close
        </button>
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;
